File: Samples/Beyond.NET.Sample.Android/app/src/androidTest/java/com/example/beyondnetsampleandroid/SystemSecurityCryptographyTests.kt
// NOTE: This requires OpenSSL libraries (.so files).
// Right now, libcrypto.so and libssl.so included in this repo are sourced from https://github.com/Sharm/Prebuilt-OpenSSL-Android which is ancient.
// TODO: Find a better way to include OpenSSL

package com.example.beyondnetsampleandroid

import androidx.test.ext.junit.runners.AndroidJUnit4

import org.junit.Test
import org.junit.runner.RunWith

import org.junit.Assert.*

import com.example.beyondnetsampleandroid.dn.*

@RunWith(AndroidJUnit4::class)
class SystemSecurityCryptographyTests {
    @Test
    fun testAES() {
        val unencryptedData = "Hello 🌎!"

        val encryptResult = encrypt(unencryptedData)
        val encryptedData = encryptResult.encryptedData
        val iv = encryptResult.iv
        val key = encryptResult.key

        assertNotEquals(encryptedData, "")

        val decryptedData = decrypt(encryptedData, iv, key)

        assertEquals(decryptedData, unencryptedData)
    }

    class EncryptRet(val encryptedData: String, val iv: DNArray<System_Byte>, val key: DNArray<System_Byte>)

    private fun encrypt(data: String): EncryptRet {
        val aes = System_Security_Cryptography_Aes.create()

        aes.generateIV()
        aes.generateKey()

        val iv = aes.iV_get()
        val key = aes.key_get()

        val encryptor = aes.createEncryptor()
        val memoryStream = System_IO_MemoryStream()

        val cryptoStream = System_Security_Cryptography_CryptoStream(memoryStream, encryptor, System_Security_Cryptography_CryptoStreamMode.WRITE)
        val streamWriter = System_IO_StreamWriter(cryptoStream)

        streamWriter.write(data.toDotNETString())
        cryptoStream.flush()

        streamWriter.dispose()
        cryptoStream.dispose()
        memoryStream.dispose()
        aes.dispose()

        val encryptedData = memoryStream.toArray()

        val base64EncryptedDataDN = System_Convert.toBase64String(encryptedData)
        val base64EncryptedData = base64EncryptedDataDN.toKString()

        return EncryptRet(base64EncryptedData, iv, key)
    }

    private fun decrypt(data: String, iv: DNArray<System_Byte>, key: DNArray<System_Byte>): String {
        val buffer = System_Convert.fromBase64String(data.toDotNETString())

        val aes = System_Security_Cryptography_Aes.create()
        aes.iV_set(iv)
        aes.key_set(key)

        val decryptor = aes.createDecryptor()
        val memoryStream = System_IO_MemoryStream(buffer)

        val cryptoStream = System_Security_Cryptography_CryptoStream(memoryStream, decryptor, System_Security_Cryptography_CryptoStreamMode.READ)
        val streamReader = System_IO_StreamReader(cryptoStream)

        cryptoStream.flush()

        val decryptedDataDN = streamReader.readToEnd()
        val decryptedData = decryptedDataDN.toKString()

        cryptoStream.dispose()
        memoryStream.dispose()
        aes.dispose()
        streamReader.dispose()

        return decryptedData
    }
}

File: Samples/Beyond.NET.Sample.Swift/Tests/System/SystemSecurityCryptographyTests.swift
import XCTest
import BeyondDotNETSampleKit

final class SystemSecurityCryptographyTests: XCTestCase {
    override class func setUp() {
        Self.sharedSetUp()
    }
    
    override class func tearDown() {
        Self.sharedTearDown()
    }
    
    func testAES() throws {
        let unencryptedData = "Hello 🌎!"
        
        let encryptResult = try encrypt(data: unencryptedData)
        let encryptedData = encryptResult.encryptedData
        let iv = encryptResult.iv
        let key = encryptResult.key
        
        XCTAssertNotEqual(encryptedData, "")
        
        let decryptedData = try decrypt(data: encryptedData,
                                        iv: iv,
                                        key: key)
        
        XCTAssertEqual(decryptedData, unencryptedData)
    }
}

fileprivate extension SystemSecurityCryptographyTests {
    func encrypt(data: String) throws -> (encryptedData: String, 
                                          iv: DNArray<System.Byte>,
                                          key: DNArray<System.Byte>) {
        let aes = try System.Security.Cryptography.Aes.create()
        
        try aes.generateIV()
        try aes.generateKey()
        
        let iv = try aes.iV
        let key = try aes.key
        
        let encryptor = try aes.createEncryptor()
        let memoryStream = try System.IO.MemoryStream()
        
        let cryptoStream = try System.Security.Cryptography.CryptoStream(memoryStream,
                                                                         encryptor,
                                                                         .write)
        
        let streamWriter = try System.IO.StreamWriter(cryptoStream)
        
        try streamWriter.write(data.dotNETString())
        
        try cryptoStream.flush()
        
        try streamWriter.dispose()
        try cryptoStream.dispose()
        try memoryStream.dispose()
        try aes.dispose()
        
        let encryptedData = try memoryStream.toArray()
        
        let base64EncryptedDataDN = try System.Convert.toBase64String(encryptedData)
        let base64EncryptedData = base64EncryptedDataDN.string()
        
        return (encryptedData: base64EncryptedData,
                iv: iv,
                key: key)
    }
    
    func decrypt(data: String,
                 iv: DNArray<System.Byte>,
                 key: DNArray<System.Byte>) throws -> String {
        let buffer = try System.Convert.fromBase64String(data.dotNETString())
        
        let aes = try System.Security.Cryptography.Aes.create()
        try aes.iV_set(iv)
        try aes.key_set(key)
        
        let decryptor = try aes.createDecryptor()
        let memoryStream = try System.IO.MemoryStream(buffer)
        
        let cryptoStream = try System.Security.Cryptography.CryptoStream(memoryStream,
                                                                         decryptor,
                                                                         .read)
        
        let streamReader = try System.IO.StreamReader(cryptoStream)
        
        try cryptoStream.flush()
        
        let decryptedDataDN = try streamReader.readToEnd()
        let decryptedData = decryptedDataDN.string()
        
        try cryptoStream.dispose()
        try memoryStream.dispose()
        try aes.dispose()
        try streamReader.dispose()
        
        return decryptedData
    }
}
